### PR TITLE
feat(builder): delete cache if the cache is disabled

### PR DESCRIPTION
**Pull request by Joshua-Anderson**
cc @jeroenvisser101

If `DEIS_DISABLE_CACHE` is set and the cache file exists, delete the cache.

### CODE REVIEW

File: pkg/gitreceive/build.go [hunk 1]
```diff
@@ -112,6 +110,20 @@ func build(
 		}
 	}
 
+	_, disableCaching := appConf.Values["DEIS_DISABLE_CACHE"]
+	slugBuilderInfo := NewSlugBuilderInfo(appName, gitSha.Short(), disableCaching)
+
+	if slugBuilderInfo.DisableCaching() {
+		log.Debug("caching disabled for app %s", appName)
+		// If cache file exists, delete it
+		if _, err := storageDriver.Stat(context.Background(), slugBuilderInfo.CacheKey()); err == nil {
```
Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: Not sure, but I think these operations cost money, although I might be wrong.

File: pkg/gitreceive/build.go [hunk 2]
```diff
@@ -178,14 +190,19 @@ func build(
 		)
 	} else {
 		buildPodName = slugBuilderPodName(appName, gitSha.Short())
+
+		cacheKey := ""
+		if !slugBuilderInfo.DisableCaching() {
+			cacheKey = slugBuilderInfo.CacheKey()
+		}
```
Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: Why did you add this exactly? When the cache is disabled, the cache key should still be a correct 'cache key', right? It should just be disabled, which it is.

**Comment by jeroenvisser101** [comment]
Few remarks, but overal, this LGTM.

File: pkg/gitreceive/build.go [hunk 3]
```diff
@@ -112,6 +110,20 @@ func build(
 		}
 	}
 
+	_, disableCaching := appConf.Values["DEIS_DISABLE_CACHE"]
```
Comment (kmala) [inline pkg/gitreceive/build.go]: this would mean user has to do config:set/unset DEIS_DISABLE_CACHE to disable/enable the cache.
Can we check the value instead?

Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: iirc, the application is now checking if it's 'set', which would mean this logic is correct. We could still change it though, although having `DEIS_DISABLE_CACHE=0` makes no sense to me and I'd rather see it being unset.

Comment (kmala) [inline pkg/gitreceive/build.go]: i think having `DEIS_DISABLE_CACHE=true/false` is better than set/unset

Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: I think my preference would be to do the same as `DEIS_KUBERNETES_DEPLOYMENTS` (I'm not sure how it's done there, but if we do it the same, we're consistent). Also, if you feel strongly about this, we might want to change both.

File: pkg/gitreceive/build.go (hunk 2 above)
Comment (Joshua-Anderson) [inline pkg/gitreceive/build.go]: I removed some logic in `slugbuilderPod`, so it only adds `cacheKey` to the env if it's non-empty, instead of checking the map.

Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: ah, seems alright then 👍

File: pkg/gitreceive/build.go (hunk 1 above)
Comment (Joshua-Anderson) [inline pkg/gitreceive/build.go]: They do cost money, but it's `$0.01 / 10K` per read API request on s3 at least, so I'd consider it negligible.

So it's much cheaper to check before attempting to delete, which costs `$0.01 / 1K`

Comment (Joshua-Anderson) [inline pkg/gitreceive/build.go]: I was incorrect, deleting a object on S3 is free. That being said, I suspect it's better practice to check if the file exists *before* trying to delete it, but that's me.

File: pkg/gitreceive/build.go (hunk 3 above)
Comment (Joshua-Anderson) [inline pkg/gitreceive/build.go]: `DEIS_KUBERNETES_DEPLOYMENTS` worked similarly to this, but it's no longer in use, deployments having become the default.

File: pkg/gitreceive/build.go (hunk 1 above)
Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: > They do cost money, but it's $0.004 / 10K per read API request on s3 at least, so I'd consider it negligible.

Totally, that's less than 1 cent per 20k builds, seems okay to me.

File: pkg/gitreceive/build.go (hunk 3 above)
Comment (jeroenvisser101) [inline pkg/gitreceive/build.go]: Any other uses of these feature flags? Any pros/cons of using true/false vs. set/unset?

**Comment by mboersma** [comment]
Set/unset is ok with me, since we did create a precedent for it with the deployments feature flag.
